pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,426 @@
+	return nil
+}
+
+func (s *ServiceController) deleteExternalLoadBalancer(service *api.Service) error {
+	// This is only needed because not all delete load balancer implementations
+	// are currently idempotent to the LB not existing.
+	if exists, err := s.balancer.TCPLoadBalancerExists(s.loadBalancerName(service), s.zone.Region); err != nil {
+		return err
+	} else if !exists {
+		return nil
+	}
+
+	if err := s.balancer.DeleteTCPLoadBalancer(s.loadBalancerName(service), s.zone.Region); err != nil {
+		return err
+	}
+	return nil
+}
+
+// listKeys implements the interface required by DeltaFIFO to list the keys we
+// already know about.
+func (s *ServiceController) listKeys() []string {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	keys := make([]string, len(s.serviceMap))
+	i := 0
+	for k := range s.serviceMap {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+func (s *ServiceController) getService(serviceName string) (*api.Service, bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	info, ok := s.serviceMap[serviceName]
+	return info, ok
+}
+
+func (s *ServiceController) setService(serviceName string, info *api.Service) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.serviceMap[serviceName] = info
+}
+
+func (s *ServiceController) deleteService(serviceName string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	delete(s.serviceMap, serviceName)
+}
+
+func needsUpdate(oldService *api.Service, newService *api.Service) bool {
+	if !oldService.Spec.CreateExternalLoadBalancer && !newService.Spec.CreateExternalLoadBalancer {
+		return false
+	}
+	if oldService.Spec.CreateExternalLoadBalancer != newService.Spec.CreateExternalLoadBalancer {
+		return true
+	}
+	if !portsEqual(oldService, newService) || oldService.Spec.SessionAffinity != newService.Spec.SessionAffinity {
+		return true
+	}
+	if len(oldService.Spec.PublicIPs) != len(newService.Spec.PublicIPs) {
+		return true
+	}
+	for i := range oldService.Spec.PublicIPs {
+		if oldService.Spec.PublicIPs[i] != newService.Spec.PublicIPs[i] {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *ServiceController) loadBalancerName(service *api.Service) string {
+	return s.clusterName + "-" + service.Namespace + "-" + service.Name
+}
+
+// TODO: Deduplicate this with the copy in pkg/registry/service/rest.go.
+func getTCPPorts(service *api.Service) ([]int, error) {
+	ports := []int{}
+	for i := range service.Spec.Ports {
+		// TODO: Support UDP. Remove the check from pkg/registry/service/rest.go
+		// once it's supported.
+		sp := &service.Spec.Ports[i]
+		if sp.Protocol != api.ProtocolTCP {
+			return nil, fmt.Errorf("external load balancers for non TCP services are not currently supported.")
+		}
+		ports = append(ports, sp.Port)
+	}
+	return ports, nil
+}
+
+func portsEqual(x, y *api.Service) bool {
+	xPorts, err := getTCPPorts(x)
+	if err != nil {
+		return false
+	}
+	yPorts, err := getTCPPorts(y)
+	if err != nil {
+		return false
+	}
+	if len(xPorts) != len(yPorts) {
+		return false
+	}
+	for i := range xPorts {
```

Comment:
I just pulled this code from the REST handler, but that's a fair point, since it doesn't look like our validation functions require them to be sorted.

Since go's sorting functions are in-place and I'd rather not modify the user-specified order, I just used a map for checking the equality.